File: app/src/main/java/bj/vinylbrowser/main/MainModule.java
package bj.vinylbrowser.main;

import android.content.Context;

import bj.vinylbrowser.di.scopes.ActivityScope;
import bj.vinylbrowser.main.panel.YouTubeListFragment;
import bj.vinylbrowser.main.panel.YouTubePlayerEpxController;
import bj.vinylbrowser.main.panel.YouTubePlayerHolder;
import bj.vinylbrowser.main.panel.YouTubePlayerPresenter;
import dagger.Module;
import dagger.Provides;

/**
 * Created by Josh Laird on 01/06/2017.
 */
@Module
public class MainModule
{
    private final MainContract.View mView;

    public MainModule(MainContract.View view)
    {
        mView = view;
    }

    @Provides
    @ActivityScope
    protected MainContract.View provideView()
    {
        return mView;
    }

    @Provides
    @ActivityScope
    protected YouTubePlayerHolder provideYoutubePlayerHolder(Context context)
    {
        return new YouTubePlayerHolder(context);
    }

    @Provides
    @ActivityScope
    protected MainPresenter provideMainPresenter()
    {
        return new MainPresenter(mView);
    }

    @Provides
    @ActivityScope
    protected YouTubeListFragment provideYouTubeListFragment()
    {
        return new YouTubeListFragment();
    }

    @Provides
    protected RouterAttacher provideRouterAttacher()
    {
        return new RouterAttacher();
    }

    @Provides
    @ActivityScope
    protected YouTubePlayerEpxController provideYouTubePlayerEpxController(Context context, YouTubePlayerHolder youTubePlayerHolder, MainPresenter mainPresenter)
    {
        return new YouTubePlayerEpxController(context, youTubePlayerHolder, mainPresenter);
    }

    @Provides
    @ActivityScope
    protected YouTubePlayerPresenter provideYoutubePlayerPresenter(Context context, YouTubePlayerEpxController controller, YouTubePlayerHolder youTubePlayerHolder)
    {
        return new YouTubePlayerPresenter(context, controller, youTubePlayerHolder);
    }
}

File: app/src/main/java/bj/vinylbrowser/main/MainModule.kt
package bj.vinylbrowser.main

import bj.vinylbrowser.di.scopes.ActivityScope
import dagger.Module
import dagger.Provides

/**
 * Created by Josh Laird on 29/05/2017.
 *
 * Method and class must be declared open for DaggerMock
 */
@Module
open class MainModule(mainActivity: MainActivity) {
    lateinit var mView: MainContract.View

    fun FirstModule(view: MainContract.View) {
        mView = view
    }

    @Provides
    @ActivityScope
    open protected fun provideView(): MainContract.View {
        return mView
    }

//    @Provides
//    @ActivityScope
//    open protected fun provideNavigationDrawerBuilder(context: Context, sharedPrefsManager: SharedPrefsManager, daoManager: DaoManager): NavigationDrawerBuilder {
//        return NavigationDrawerBuilder(context, sharedPrefsManager, daoManager)
//    }
//
//    @Provides
//    @ActivityScope
//    open protected fun compositeDisposable(): CompositeDisposable {
//        return CompositeDisposable()
//    }

//    @Provides
//    @ActivityScope
//    protected fun providesFirstPresenter(context: Context, discogsInteractor: DiscogsInteractor, mySchedulerProvider: MySchedulerProvider,
//                                        builder: NavigationDrawerBuilder, controller: MainEpxController, sharedPrefsManager: SharedPrefsManager,
//                                        log: LogWrapper, daoManager: DaoManager, tracker: AnalyticsTracker): MainPresenter {
//        return MainPresenter(context, mView, discogsInteractor, mySchedulerProvider, builder, controller, sharedPrefsManager, log, daoManager, tracker)
//    }
}